_studio/shared/umc/codec/mpeg2_dec/src/umc_mpeg2_decoder.cpp

```diff
@@ -0,0 +1,1147 @@
+            info.UpdateReferences(m_dpb);
+        }
+
+        return UMC::UMC_ERR_NOT_ENOUGH_DATA;
+    }
+
+    // Decode a bitstream header NAL unit
+    UMC::Status MPEG2Decoder::DecodeHeaders(RawUnit & unit)
+    {
+        UMC::Status umcRes = UMC::UMC_OK;
+
+        MPEG2HeadersBitstream bitStream;
+
+        MPEG2_UNIT_TYPE type = (MPEG2_UNIT_TYPE)-1;
+
+        try
+        {
+            bitStream.Reset(unit.begin + prefix_size, unit.end - unit.begin - prefix_size); // "+ prefix_size" is to skip start code
+
+            type = bitStream.GetUnitType();
+
+            switch(type)
+            {
+            case SEQUENCE_HEADER:
+                {
+                    size_t size = unit.end - unit.begin;
+                    auto seqHdr = std::make_shared<MPEG2SequenceHeader>();
+                    auto seqExtHdr = std::make_shared<MPEG2SequenceExtension>();
+                    bitStream.GetSequenceHeader(*seqHdr.get());
+
+                    uint8_t* seqExtBegin = RawHeaderIterator::FindStartCode(unit.begin + prefix_size + bitStream.BytesDecoded(), unit.end);
+                    if (!seqExtBegin)
+                        return UMC::UMC_ERR_INVALID_STREAM;
+
+                    bitStream.Reset(seqExtBegin + prefix_size, unit.end - seqExtBegin);
+                    bitStream.SkipBits(8 + 4); // skip unit and extension types
+                    bitStream.GetSequenceExtension(*seqExtHdr.get());
+
+                    // Save headers
+                    m_currHeaders.seqHdr = seqHdr;
+                    m_currHeaders.seqExtHdr = seqExtHdr;
+
+                    // Reset 'dependent' headers
+                    m_currHeaders.displayExtHdr.reset();
+                    m_currHeaders.qmatrix.reset();
+                    m_currHeaders.picHdr.reset();
+                    m_currHeaders.picExtHdr.reset();
+
+                    m_rawHeaders.Resize(unit.end - unit.begin);
+                    std::copy(unit.begin, unit.end, m_rawHeaders.GetPointer());
+                }
+                break;
+            case PICTURE_HEADER:
+                {
+                    size_t size = unit.end - unit.begin;
+                    auto picHdr = std::make_shared<MPEG2PictureHeader>();
+                    auto picExtHdr = std::make_shared<MPEG2PictureCodingExtension>();
+
+                    bitStream.GetPictureHeader(*picHdr.get());
+
+                    uint8_t* picExtBegin = RawHeaderIterator::FindStartCode(unit.begin + prefix_size + bitStream.BytesDecoded(), unit.end);
+                    if (!picExtBegin)
+                        return UMC::UMC_ERR_INVALID_STREAM;
+
+                    bitStream.Reset(picExtBegin + prefix_size, unit.end - picExtBegin);
+                    bitStream.SkipBits(8 + 4); // skip unit and extension types
+                    bitStream.GetPictureExtensionHeader(*picExtHdr.get());
+
+                    // Save headers
+                    m_currHeaders.picHdr = picHdr;
```

Comment:
use move assignment for shared_ptr, it has quite expensive copy constructor because of atomic counters in control block

Reply:
Ok. Thanks